app/gui2/shared/ast/parse.ts

```diff
@@ -591,11 +596,121 @@ function resync(
+    parent.syncToCode(goodAst.code())
   }
 
   console.warn(
     `repair: Replaced ${parentsOfBadSubtrees.size} subtrees with their reparsed equivalents.`,
     parentsOfBadSubtrees,
   )
 }
+
+function hashString(s: string) {
+  return hash(s)
+}
+
+function hashSubtree(ast: Ast, hashesOut: Map<string, Ast[]>) {
+  let content = ''
+  content += ast.typeName + ':'
+  for (const child of ast.concreteChildren()) {
+    content += child.whitespace ?? ' '
+    if (isTokenId(child.node)) {
+      content += 'Token:' + hashString(ast.module.getToken(child.node).code())
+    } else {
+      content += hashSubtree(ast.module.checkedGet(child.node), hashesOut)
+    }
+  }
+  const astHash = hashString(content)
+  map.setIfUndefined(hashesOut, astHash, (): Ast[] => []).unshift(ast)
+  return astHash
+}
+
+function hashTree(root: Ast) {
+  const hashes = new Map<string, Ast[]>()
+  const rootHash = hashSubtree(root, hashes)
+  return { root: rootHash, hashes }
+}
+
+function rawBlockToInline(tree: RawAst.Tree.Tree) {
+  if (tree.type !== RawAst.Tree.Type.BodyBlock) return tree
+  const statements = Array.from(tree.statements)
+  if (statements.length !== 1) return tree
+  const expression = statements[0]!.expression
+  if (expression === undefined) return tree
+  return expression
+}
+
+function analyzeCode(code: string, asBlock: boolean, module: MutableModule) {
+  const rawParsedBlock = parseEnso(code)
+  const rawParsed = asBlock ? rawParsedBlock : rawBlockToInline(rawParsedBlock)
+  // NOTE: We could do the initial analysis much more efficiently by operating on the raw tree; we just need to separate
+  // the logic for mapping raw types to `Ast` from the actual construction of the `Ast` types.
+  const parsed = abstract(module, rawParsed, code)
+  const newNodesBySpan = parsed.spans.nodes
+  const newSpans = new Map<AstId, SourceRange>()
+  for (const [key, asts] of parsed.spans.nodes) {
+    for (const ast of asts) newSpans.set(ast.id, sourceRangeFromKey(key))
+  }
+  return { parsedRoot: parsed.root, newSpans, newNodesBySpan, rawParsed }
+}
+
+function findIdenticalSubtrees(ast1: Ast, ast2: Ast) {
+  const ast1ByContent = hashTree(ast1)
+  const ast2ByContent = hashTree(ast2)
+  const ast1MatchedNodes = new Set<AstId>()
+  for (const [key, asts] of ast1ByContent.hashes) {
+    if (ast2ByContent.hashes.has(key)) {
+      for (const ast of asts) ast1MatchedNodes.add(ast.id)
+    }
+  }
+  const ast1MatchedSubtrees = new Set(subtreeRoots(ast1.module, ast1MatchedNodes))
+  const equivalences = new Array<readonly [AstId, AstId]>()
+  for (const [key, asts] of ast1ByContent.hashes) {
+    const ast1Roots = asts.filter((ast) => ast1MatchedSubtrees.has(ast.id))
+    const ast2Equivalents = ast2ByContent.hashes.get(key)
+    for (;;) {
+      const ast1Root = ast1Roots.shift()
+      if (!ast1Root) break
+      const ast2Root = ast2Equivalents!.shift()
+      if (!ast2Root) break
+      equivalences.push([ast1Root.id, ast2Root.id])
+    }
```

Comment:
I think we are prone to hash collisions, aren't we?

Of course, it's not easy to check actual content without falling to quadratic time complexity. Perhaps it should be done in "hashSubtree" where are checking would start at token level (and if any hashed string collides with a different string, we could add a sort of "version" to it).

Reply:
The simplest solution is to use a hash function with an astronomically-small probability of collision. `hash-sum` is not a good choice; it was just convenient for an initial working version. Any reasonable 128-bit hash (i.e. without a collision rate much higher than the birthday-paradox rate) would be completely safe for our purposes (non-adversarial comparisons of sub-gigabyte files).